test/e2e/logforwarding/fluent/fluent_test.go

```diff
@@ -0,0 +1,154 @@
+package fluent_test
+
+import (
+	"context"
+	"fmt"
+
+	. "github.com/onsi/ginkgo"
+	loggingv1 "github.com/openshift/cluster-logging-operator/pkg/apis/logging/v1"
+	"github.com/openshift/cluster-logging-operator/test"
+	"github.com/openshift/cluster-logging-operator/test/client"
+	"github.com/openshift/cluster-logging-operator/test/helpers/cmd"
+	"github.com/openshift/cluster-logging-operator/test/helpers/fluentd"
+	. "github.com/openshift/cluster-logging-operator/test/matchers"
+	"github.com/openshift/cluster-logging-operator/test/runtime"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const message = "My life is my message"
+
+var (
+	// Create/delete clusterlogging instance around entire suite.
+	cl = runtime.NewClusterLogging()
+	_  = BeforeSuite(func() { ExpectOK(client.Get().Recreate(cl)) })
+	_  = AfterSuite(func() { _ = client.Get().Remove(cl) })
+)
+
+// Extend client.Test for tests using logGenerator and receiver.
+type Test struct {
+	*client.Test
+	receiver     *fluentd.Receiver
+	logGenerator *corev1.Pod
+	group        test.FailGroup // Run cluster operations in parallel
+}
+
+func (t *Test) Reader(name string) *cmd.Reader {
+	return t.receiver.Sources[name].TailReader()
+}
+
+func NewTest(message string) *Test {
+	ct := client.NewTest()
+	t := &Test{
+		Test:         ct,
+		receiver:     fluentd.NewReceiver(ct.NS.Name, "receiver"),
+		logGenerator: runtime.NewLogGenerator(ct.NS.Name, "log-generator", 10000, 0, message),
+	}
+	t.group.Go(func() { ExpectOK(t.Create(t.logGenerator)) })
+	return t
+}
+
+func (t *Test) Close() {
+	defer t.Test.Close()
+	t.group.Wait()
+}
+
+var _ = Describe("[ClusterLogForwarder]", func() {
+	var t *Test
+	BeforeEach(func() { t = NewTest(message) })
+	AfterEach(func() { t.Close() })
+
+	Context("with app/infra/audit receiver", func() {
+		BeforeEach(func() {
+			t.receiver.AddSource("application", "forward", 24224)
+			t.receiver.AddSource("infrastructure", "forward", 24225)
+			t.receiver.AddSource("audit", "forward", 24226)
+			t.group.Go(func() { ExpectOK(t.receiver.Create(t.Client)) })
+		})
+
+		It("forwards application logs only", func() {
+			clf := runtime.NewClusterLogForwarder()
+			addPipeline(clf, t.receiver.Sources["application"])
+			t.group.Go(func() { ExpectOK(t.Recreate(clf)) })
+			// Use ctx to let empty checks run until we get the expected lines.
+			ctx, cancel := context.WithCancel(context.Background())
+			t.group.Go(func() {
+				defer cancel() // Cancel goroutines verifying empty when we get our logs.
```

Comment:
Is this something that can be automatically done for us via a helper method?

Reply:
We could mke the WithCancel line shorter with a terse function name, but I think that's a bad case of choosing terse over explicit. 

Possibly you could persuade me to take a different approach entirely - the idea here is that we are expecting logs from some sources, and not from others. Therefore we wait (forever) on the sources where we expect nothing, and we cancel those waits when we've got everything we expected from the other sources.

A simpler alternative would be to wait with a fixed timeout where we expect nothing, then no need for cancellation. I don't like fixed timeouts because they are always too long or too short, and that's one of the reasons our tests currently spend a lot of time sleeping for nothing. Using the wait time for expected results seems like a nice natural timeout based real conditions when the test runs.

Reply:
I'm not saying we should not do this or this is not correct.  I'm trying to emphasize the "why"?  Why am I doing this and if I'm writing a test is it going to be obvious that I forgot to add it.  Can the framework/test design make it more obvious what to  do and why by either:
* making a name that makes it explicit
* Adding a function that "does it for me"

Reply:
The goroutine that starts at line 75 is going to block until it reads some expected log output (or the read times out).
Following this (not shown above) are two goroutines that will block reading on files that shoud *NOT* get any data. I pass this context to those Read() methods. If Read()s succeed that's an error - I wasn't expecting data there.
Once the first groutine succeeds, it cancels the others, otherwise the test will be stuck for some overly-long timeout waiting for data that won't come.
I'm not sure how to express that more clearly - context.Context is the standard way to cancel goroutines if their work is no longer needed. "cancel()" is the name the context library uses.